### PR TITLE
status-indicator: Display external service syncing errors

### PR DESCRIPTION
**Pull request by mrnugget**
This is the revised version of the (never merged) PR #4780 and the first attempt at fixing #5146.

**TL;DR:** This is step 1/n. It shows `repo-updater` errors in the UI. They look raw, but I think that's fine, because we finally show errors and I'm now also happy with the underlying implementation and think that we can iterate on the details (see later sections)

What this PR does it to make the _errors_ visible that `repo-updater` encounters when syncing repositories from external services. These errors can be produced when trying to construct an external service from the existing configuration or when talking to the external service to list the available repositories for sync.

These errors then get shown in the UI, like this:

Screen Shot 2019-08-21 at 17 05 29

### Implementation

The implementation in _this_ PR is based on #4780:

* It adds a new type of status message, a `SyncErrorStatusMessage`
* This status message is returned by `repo-updater` in its `/status-messages` endpoint when a sync error has been produced in the last sync run
* Each sync run sets or resets the current sync error, meaning: fixing an external service configuration results in the message disappearing with the next successful sync (triggered on save). But that also means that errors that pop up long after an external service has been created are displayed (network errors, invalid tokens, ...)

In addition to that it also contains the suggestions made by @felixfbecker and @tsenart:

* The `StatusMessage` type in the GraphQL schema has been turned into a union type. I like this approach a lot more, because it not only gives us type-safety in TypeScript, but (at the cost of verbosity) also gives us better type-safety in `graphqlbackend` and `repo-updater`.
* There are no union types in Go, which makes the definition of `StatusMessage` in `repoupdater/protocol` a bit unwiedly. But that's one of the possible I discussed with @tsenart and I like it a lot more than the untyped list of key-value pairs.
* The sync errors are now styled with `alert-danger` which makes them pop out a lot more as errors

### Open issues, drawbacks & thoughts for enhancements

1. **The error messages are still _raw_**. But now that I'm happier with the underlying implementation I agree with @nicksnyder's assessment that showing _anything_ in case there's an error is better than showing nothing. And It's actually not that easy to untangle layers and layers of errors produced by (possibly multiple) external services and turn them into something meaningful and actionable. But I will use the rest of this iteration to hack on this problem. 
**Idea**: I prototyped to define a set of errors that is common across external services (bad credentials, connection error, rate limit, ...) and then wrap the errors produced by each external service into one of these custom errors. In the `repo-updater` endpoint we could then detect these errors and replace their messages with something meaningful. The drawback of that is that it's hard to define this common set so it makes sense for all external services and wrapping each the errors produced by each one requires knowledge of each external service's idosyncracies. That's a lot of work, **but** we could approach it step by step and wrap/beautify these errors as they arise.
2. **This does not (yet?) show the _warnings_ that `repo-updater` produces.** Example of such a warning is "repository not found" which is produced when we try to fetch the configured repositories but can't find one of them. This warning is only written to the log and never returned as an error, because we want to 
3. **This does not (yet?) show errors/warnings that `gitserver` runs into when cloning/fetching/cleaning repositories**. Until now I haven't looked into making `gitserver` errors visible, but that's certainly a large part of what would make fixing #5146 valuable

Especially **2** and **3** make me question whether the current "architecture" will hold up: if we only care about errors, it's easy to remember the last error and reset it when everything's fine again. But what do we do when we want to display _ephemeral data_, such as warnings? Where and how do we collect it? (Theoretically we _could_ change our function's return signature from `<val>, error` to `<val>, warningstring, error`, but that's ... ugly). I keep thinking that we need to introduce something (an `ErrorReporter`?) that other functions/methods could call to report an issue. That would be far easier than wrapping, passing and unwrapping errors along the call chain in the syncer, with the possible added cost of introducing a singleton. And, still: what do we do with warnings?

Happy about any input/feedback/ideas!

### Test Plan

Test plan: manual testing, go test, cd web && yarn test --runInBand StatusMessagesNavItem.test.tsx

### CODE REVIEW

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 1]
```diff
@@ -88,9 +108,31 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
...
-    private renderMessage(message: GQL.IStatusMessage): JSX.Element | null {
-        switch (message.type) {
-            case GQL.StatusMessageType.CLONING:
+    private renderSyncErrorMessage(message: GQL.ISyncErrorStatusMessage): JSX.Element | null {
+        const displayName = message.externalServiceDisplayName
+        const extSvcID = message.externalServiceId
+
+        const title = `Syncing external service "${displayName}" failed:`
+        const linkTo = `/site-admin/external-services/${extSvcID}`
+        const linkText = `Edit "${displayName}"...`
```
Comment (lguychard) [inline web/src/nav/StatusMessagesNavItem.tsx]: Personal preference: I don't think the `...` is necessary.

```suggestion
        const linkText = `Edit "${displayName}"`
```

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 2]
```diff
@@ -88,9 +108,31 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
         this.subscriptions.unsubscribe()
     }
 
-    private renderMessage(message: GQL.IStatusMessage): JSX.Element | null {
-        switch (message.type) {
-            case GQL.StatusMessageType.CLONING:
+    private renderSyncErrorMessage(message: GQL.ISyncErrorStatusMessage): JSX.Element | null {
```
Comment (lguychard) [inline web/src/nav/StatusMessagesNavItem.tsx]: This is terser, and avoids declaring `displayName` and `extSvcID`.

```suggestion
    private renderSyncErrorMessage({ message, externalServiceDisplayName, externalServiceId}: GQL.ISyncErrorStatusMessage): JSX.Element | null {
```

Overall though, I feel like this doesn't need to be a separate function and could just be inlined.

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 3]
```diff
@@ -34,22 +44,32 @@ interface StatusMessageEntryProps {
...
-        )}
-    </div>
-)
+const StatusMessagesNavItemEntry: React.FunctionComponent<StatusMessageEntryProps> = props => {
+    let className = 'status-messages-nav-item__entry'
+    if (props.alert) {
+        className = `${className} alert ${props.alert} mb-0`
+    }
+    return (
+        <div key={props.text} className={className}>
```
Comment (lguychard) [inline web/src/nav/StatusMessagesNavItem.tsx]: You could use `classNames()`:

```suggestion
        <div key={props.text} className={classNames('status-messages-nav-item__entry', props.alert && `alert ${props.alert} mb-0`)}>
```

**Comment by mrnugget** [comment]
Thanks for the great suggestions, @lguychard! Wil incorporate them today!

> It looks a bit odd to me that in the attached screenshot, the 'Repositories Cloning' message is still shown. I feel like the display of errors & cloning messages should be mutually exclusive. Thoughts?

I don't think they should be mutually exclusive. There's multiple cases where it's useful to have this:
* You've added an external service for GitHub that clones ~10k repos. Takes some time. Then you add one for Gitlab and that produces a warning because, e.g. the connection has problems. You still want to see that GitHub makes progress while you deal with the other issue.
* You've added two external services in the past. Someone added a new org with 1000 repos to one code host. Sourcegraph starts cloning. At the same time someone deleted the token you use to authenticate with the other external service. It still makes sense to show that one external service is doing fine and the other is not.

Point being: external services can be totally independent of each other and I think it makes sense to show the cloning message even if one external service produces errors. It's also an added signal: does the number of to-be-cloned repos change while the error is displayed? Or is it not? Could help debugging

File: web/src/nav/StatusMessagesNavItem.tsx (hunk 3 above)
Comment (mrnugget) [reply web/src/nav/StatusMessagesNavItem.tsx]: Ahh, good stuff!

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 4]
```diff
@@ -134,14 +195,28 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
                             <p className="alert alert-danger">{startCase(this.state.messagesOrError.message)}</p>
                         </div>
                     ) : this.state.messagesOrError.length > 0 ? (
-                        this.state.messagesOrError.map(m => this.renderMessage(m))
+                        this.state.messagesOrError.map((m, i) => {
+                            if (!isErrorLike(this.state.messagesOrError) && i < this.state.messagesOrError.length - 1) {
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply web/src/nav/StatusMessagesNavItem.tsx]: It is, but the TypeScript compiler seems to need it:

```
Property 'length' does not exist on type 'StatusMessage[] | ErrorLike'.
  Property 'length' does not exist on type 'ErrorLike'.
```

**Comment by lguychard** [comment]
> You still want to see that GitHub makes progress while you deal with the other issue.

> It still makes sense to show that one external service is doing fine and the other is not.

I understand this need, but this isn't conveyed clearly (to me) in the current implementation, mostly because the cloning message, as shown, is not specific to an external service, while the error messages are. This makes it unclear to me whether the errors are blocking the enqueued repositories from being cloned, or whether cloning of some repositories is still happening in spite of the error.

**Comment by mrnugget** [comment]
> I understand this need, but this isn't conveyed clearly (to me) in the current implementation, mostly because the cloning message, as shown, is not specific to an external service, while the error messages are.

I see. Good point. Solving this is not quite that easy, I feel, because of the second part:

>  This makes it unclear to me whether the errors are blocking the enqueued repositories from being cloned, or whether cloning of some repositories is still happening in spite of the error.

The errors _could_ block the cloning, but not necessarily.

Off the top of my head, there's two potential solutions to this:
* We need to be more specific in the cloning progress (ideally we could show the repos that are being cloned right now, or have a progress bar, or show which external service is involved, etc.). Requires a lot of added state/functionality to gitserver/repo-updater.
* We show a status-icon _per external service_. If everything's fine and an external service has been synced: green checkmark. if there's an error: red X. That way you could see at a glance which external service is syncing and which one's not. @christinaforney had this idea a while back. That could be a next step on the status indicator.

File: cmd/repo-updater/repos/syncer.go [hunk 5]
```diff
@@ -311,13 +318,39 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
 
 	srcs, err := s.sourcer(svcs...)
 	if err != nil {
+		// Only reset source error if it was non-nil. Because if `ListRepos`
+		// produces an error through multiple syncs, the state of
+		// multiSourceErr would flip-flop between here and the `ListRepos` call
+		// further down.
+		s.SetOrResetMultiSourceErr(err)
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/syncer.go]: github won't let me comment on it, but what if ListExternalServices fails? Maybe we should just do this at the layer we called `sourced`?

File: pkg/repoupdater/protocol/repoupdater.go [hunk 6]
```diff
@@ -177,12 +177,25 @@ type ExternalServiceSyncResult struct {
 type StatusMessageType string
 
 const (
-	CloningStatusMessage StatusMessageType = "CLONING"
+	Cloning   StatusMessageType = "CLONING"
+	SyncError StatusMessageType = "SYNCERROR"
```
Comment (keegancsmith) [inline pkg/repoupdater/protocol/repoupdater.go]: you can use go's `stringer` command to generate a nice enum

File: pkg/repoupdater/protocol/repoupdater.go [hunk 7]
```diff
@@ -177,12 +177,25 @@ type ExternalServiceSyncResult struct {
...
+	Message                    string
+	ExternalServiceId          int64
+	ExternalServiceKind        string
+	ExternalServiceDisplayName string
+}
+
 type StatusMessage struct {
-	Message string            `json:"message"`
-	Type    StatusMessageType `json:"type"`
+	Type StatusMessageType `json:"type"`
```
Comment (keegancsmith) [inline pkg/repoupdater/protocol/repoupdater.go]: not sure: maybe we should just rely on the mutual exclusivity of the below pointers to specify type?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 8]
```diff
@@ -3719,19 +3719,28 @@ type ProductSubscriptionEvent {
...
     # The message of this status message
     message: String!
-    # The type.
-    type: StatusMessageType!
+    # The id of the external service that failed to sync
+    externalServiceId: String!
+    # The kind of the external service that failed to sync
+    externalServiceKind: String!
+    # The display name of the external service that failed to sync
+    externalServiceDisplayName: String!
```
[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/graphqlbackend/schema.graphql]: Yeah, in GraphQL you usually don't want to reference other resources by ID but by resolver

File: web/src/nav/StatusMessagesNavItem.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/nav/StatusMessagesNavItem.tsx]: This is because you are in a callback and referencing a property, not a constant. So by the time the callback is executed the type could be different again. You should be able to avoid this by saving `this.state.messagesOrError` to a local constant in `render()`.

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 9]
```diff
@@ -112,36 +172,38 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
...
                             <h4>Failed to load status messages:</h4>
-                            <p className="alert alert-danger">{startCase(this.state.messagesOrError.message)}</p>
+                            <p>{startCase(this.state.messagesOrError.message)}</p>
                         </div>
                     ) : this.state.messagesOrError.length > 0 ? (
-                        this.state.messagesOrError.map(m => this.renderMessage(m))
+                        this.state.messagesOrError.map((m, i) => {
+                            if (!isErrorLike(this.state.messagesOrError) && i < this.state.messagesOrError.length - 1) {
+                                return (
+                                    <div>
```
Comment (felixfbecker) [inline web/src/nav/StatusMessagesNavItem.tsx]: Do you need a `div` here or would a fragment `<>` be enough?

You could probably avoid the two different `returns` and instead just conditionally render the divider with `&&`

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 10]
```diff
@@ -34,22 +45,29 @@ interface StatusMessageEntryProps {
     showLink?: boolean
     linkTo: string
     linkText: string
+    alert?: string
```
Comment (felixfbecker) [inline web/src/nav/StatusMessagesNavItem.tsx]:
```suggestion
    alert?: 'danger' | 'warning' | 'info' | 'success'
```

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 11]
```diff
@@ -34,22 +45,29 @@ interface StatusMessageEntryProps {
...
+    linkOnClick: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void
 }
 
 const StatusMessagesNavItemEntry: React.FunctionComponent<StatusMessageEntryProps> = props => (
-    <div key={props.text} className="status-messages-nav-item__entry">
+    <div
+        key={props.text}
+        className={classNames('status-messages-nav-item__entry', props.alert && `alert ${props.alert} mb-0`)}
```
Comment (felixfbecker) [inline web/src/nav/StatusMessagesNavItem.tsx]:
```suggestion
        className={classNames('status-messages-nav-item__entry', props.alert && `alert alert-${props.alert} mb-0`)}
```

File: cmd/repo-updater/repos/syncer.go (hunk 5 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: `ListExternalServices` won't return a `SourceError`/`MultiSourceError`, which means it doesn't came with an external service attached to it.

I agree, though, that we should display `ListExternalServices` errors and also `UpsertRepos` errors, for that matter, since the latter was involved in a a few customer issues in the past (unique constraint errors, etc.).

That would mean that we generalize and go from a `multiSourceErr *MultiSourceError` field on `Syncer` to a `lastSyncErr error` field (which is what I had in my first internal proof-of-concept version of this). On the caller side, in `Server`, we could then differentiate between a normal `error` (without external service!) and a `MultiSourceError` (with external service!) and produce two different status messages — `SyncErrorStatusMessage` and `ListExternalServiceErrorStatusMessage`.

What do you think? Should that be part of this PR?

File: web/src/nav/StatusMessagesNavItem.tsx (hunk 9 above)
Comment (mrnugget) [reply web/src/nav/StatusMessagesNavItem.tsx]: I didn't know about fragments! Sweet! 🕺

File: web/src/nav/StatusMessagesNavItem.tsx (hunk 10 above)
Comment (mrnugget) [reply web/src/nav/StatusMessagesNavItem.tsx]: [\*Italian Chef Kiss\*](https://i.kym-cdn.com/photos/images/newsfeed/001/168/676/09e.gif)

File: cmd/repo-updater/repoupdater/server.go [hunk 12]
```diff
@@ -432,28 +432,49 @@ func (s *Server) shouldGetGithubDotComRepo(args protocol.RepoLookupArgs) bool {
...
-			Message: fmt.Sprintf("%d repositories enqueued for cloning...", notCloned),
-			Type:    protocol.CloningStatusMessage,
+			Type: protocol.Cloning,
+			Cloning: &protocol.CloningStatusMessage{
+				Message: fmt.Sprintf("%d repositories enqueued for cloning...", notCloned),
+			},
 		})
 	}
 
+	if multiErr := s.Syncer.MultiSourceError(); multiErr != nil {
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repoupdater/server.go]: That was a great suggestion. Fixed in the last commit. And while working on this I realized that we should probably make this a generic `error`... See my other comment. Curious about your thoughts.

File: cmd/repo-updater/repos/syncer.go [hunk 13]
```diff
@@ -311,13 +318,39 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
...
 	defer cancel()
 
-	return srcs.ListRepos(ctx)
+	repos, err := srcs.ListRepos(ctx)
+	s.SetOrResetMultiSourceErr(err)
+
+	return repos, err
+}
+
+func (s *Syncer) SetOrResetMultiSourceErr(err error) {
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: Good call. It was used in tests but I just changed the way the test is setup so that I can inject a test through a `repos.FakeSourcer`.

**Comment by felixfbecker** [comment]
On the styling, the dividers that don't extend to the edges of the popover look odd to me. Also, why dividers if the alerts are each in a box? Are they trying to convey some semantic groups? If yes, it is no obvious...
Also noticed "1 repositories cloning" (wrong inflection)

File: cmd/repo-updater/repos/syncer.go [hunk 14]
```diff
@@ -311,13 +318,39 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
...
+	s.multiSourceErrMu.Lock()
+
+	if multiSourceErr, ok := err.(*MultiSourceError); ok {
+		s.multiSourceErr = multiSourceErr
+	} else {
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: See the other comments wrt only caring about MultiSourceErrors for now...

**Comment by mrnugget** [comment]
> On the styling, the dividers that don't extend to the edges of the popover look odd to me. Also, why dividers if the alerts are each in a box? Are they trying to convey some semantic groups? If yes, it is no obvious...

File: cmd/repo-updater/repos/syncer.go [hunk 15]
```diff
@@ -22,6 +23,12 @@ type Syncer struct {
 	// Sourcegraph.com
 	FailFullSync bool
 
+	// LastSourcerError contains the last error returned by the Sourcer in each
+	// Sync. It's reset with each Sync and if the Sourcer produced no error,
+	// it's set to nil.
+	multiSourceErr   *MultiSourceError
+	multiSourceErrMu sync.RWMutex
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: I chose `RWMutex` because there are multiple readers and a single writer (all the requests polling `repo-updater` for status messages vs. the single `Syncer` writing to it), but I guess a normal `Mutex` will suffice. I'll update it.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: I'll give this a shot

File: pkg/repoupdater/protocol/repoupdater.go (hunk 7 above)
Comment (mrnugget) [reply pkg/repoupdater/protocol/repoupdater.go]: I've had the same thought and was on the fence, but now that you've brought it up again, I did just that and got rid of `StatusMessageType`.

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: Done! This is much better now. Thank you!

**Comment by mrnugget** [comment]
@keegancsmith @felixfbecker I've incorporated most of your suggestions, can you take another look?

Also, I'd like to tackle these two things in follow-up PRs to make review easier:

* Pass on _all_ syncing errors (not just ones related to external services), as described in [my comment here](https://github.com/sourcegraph/sourcegraph/pull/5319#discussion_r316627988)
* Tweak the design of the entries in the dropdown menu (maybe we could even pair on this a little bit @felixfbecker? If not, I'll try to come up with some options)

**Comment by felixfbecker** [comment]
As long as this targets 3.8 and we're gonna polish it before the release, I'm fine with merging and following up on the design

**Comment by mrnugget** [comment]
Just pushed some styling tweaks:
* got rid of the divider
* added a heading
* added some more `alert-*` styling

When everything is cloned:

Screen Shot 2019-08-22 at 17 31 16

With a `info` ("Cloning...") and `danger` ("error syncing") status messages:

Screen Shot 2019-08-22 at 17 28 43

I'm happy with this, but @christinaforney has some ideas, so: feel free to play with this branch and let me know whether I can merge it tomorrow or not :)

File: cmd/repo-updater/repos/syncer.go (hunk 5 above)
Comment (keegancsmith) [reply cmd/repo-updater/repos/syncer.go]: agreed this sounds good to me!

File: cmd/frontend/graphqlbackend/status_messages.go [hunk 16]
```diff
@@ -21,19 +21,58 @@ func (r *schemaResolver) StatusMessages(ctx context.Context) ([]*statusMessageRe
 		return nil, err
 	}
 
-	for _, rn := range result.Messages {
-		messages = append(messages, &statusMessageResolver{&types.StatusMessage{
-			Message: rn.Message,
-			Type:    string(rn.Type),
-		}})
+	for _, m := range result.Messages {
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/status_messages.go]: (for now) you can create a single resolver which just takes m that might be a bit clearer and less code. IE this should work

```go
func (n *statusMessageResolver) ToCloningStatusMessage() (*statusMessageResolver, bool) {
  return n, n.message.Cloning != nil
}
```

**Comment by christinaforney** [comment]
Here is what I came up with for the styles. Let me know what you think! I was hacking the DOM a bit, so anywhere you see an emoji, please use a similar icon there instead. 

I would like to see the icon indicating type of the notice:
- success ✅
- in progress/cloning ♺
- warning ⚠️
- error 🛑 (or [x] style icon)

The icon should lead the heading for proximity association, and visual representation. I used a `border-left: 6px solid #color` on the `status-messages-nav-item__entry` with the colors matching those of the alerts (see [bootstrap](https://getbootstrap.com/docs/4.0/components/alerts/)). I like this use of color, because the text stays more readable, but still applies a visual queue of the notice.

Here is a warning in light/dark:
image 
image

Success state in light/dark:
image
image

In progress and error states in light/dark:
image
image
image

**Comment by mrnugget** [comment]
I've implemented @christinaforney's ideas.

### Light theme

Screen Shot 2019-08-23 at 10 02 40
Screen Shot 2019-08-23 at 10 03 30

### Dark theme

Screen Shot 2019-08-23 at 10 02 51
Screen Shot 2019-08-23 at 10 03 53

---

We've decided that we can still iterate on this in this iteration. If someone with more design chops comes up with something better: please speak up 😄

File: cmd/frontend/graphqlbackend/status_messages.go (hunk 16 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/status_messages.go]: Thanks! Learned something new again :)

**Comment by christinaforney** [comment]
Looks great! 😁

File: web/src/nav/StatusMessagesNavItem.scss [hunk 17]
```diff
@@ -8,13 +12,25 @@
...
         }
+        &-border-success {
+            border-left: 6px solid $success;
+        }
+
+        &-border-warning {
+            border-left: 6px solid $warning;
+        }
+
+        &-border-progress {
```
Comment (felixfbecker) [inline web/src/nav/StatusMessagesNavItem.scss]: these are _modifiers_ of the same element, so they should use `--` (see BEM)

**Comment by felixfbecker** [comment]
I don't think the "line on the side" style is a bad one, but it feels a bit inconsistent to me with how we represent alerts elsewhere. I am also missing a bit of visual separation between the items, which in this design, I would solve by using the same color slighty-transparent as a background for each item. But then it really would be just an alert design, but one that is entirely different from other styles. I usually see the "line on the side" style used in design systems that prefer sharp edges & corners over rounded ones. But our design uses rounded edges everywhere. You could make the line a `div` and add a `border-radius` of course, but then it would work less as an element that implies the edges of each item (you can't mentally extend the lines anymore to draw the boundaries).

@christinaforney is there a reason why you prefer this design over our existing alert styles? For example, if it is the "box in a box" that is bothering you, maybe we could try out extending the alerts to the edges. 

👍 on icons though

File: web/src/nav/StatusMessagesNavItem.tsx [hunk 18]
```diff
@@ -28,28 +40,54 @@ export function fetchAllStatusMessages(): Observable<GQL.IStatusMessage[]> {
...
 
+type EntryType = 'warning' | 'success' | 'progress'
+
 interface StatusMessageEntryProps {
     title: string
     text: string
     showLink?: boolean
     linkTo: string
     linkText: string
+    style: EntryType
```
Comment (felixfbecker) [inline web/src/nav/StatusMessagesNavItem.tsx]: This could be confusing because `style` is associated in React with providing CSS inline styles. Perhaps `entryType` is the best

**Comment by felixfbecker** [comment]
What do you think of my idea to extend the background to the sides? E.g. https://getbootstrap.com/docs/4.3/components/list-group/#contextual-classes

File: cmd/repo-updater/repos/sources.go [hunk 19]
```diff
@@ -86,6 +86,60 @@ type Source interface {
...
+	for i, err := range es {
+		points[i] = fmt.Sprintf("* %s", err)
+	}
+
+	return fmt.Sprintf(
+		"%d errors occurred:\n\t%s\n\n",
+		len(es), strings.Join(points, "\n\t"))
+}
+
+type MultiSourceError struct {
```
Comment (tsenart) [inline cmd/repo-updater/repos/sources.go]: What makes this type necessary? I can't spot the difference between a `multierror.Error`. Sorry if this has been answered elsewhere before :-)

File: cmd/repo-updater/repos/syncer.go [hunk 20]
```diff
@@ -22,6 +23,12 @@ type Syncer struct {
 	// Sourcegraph.com
 	FailFullSync bool
 
+	// multiSourceErr contains the last error returned by the Sourcer in each
+	// Sync. It's reset with each Sync and if the Sourcer produced no error,
+	// it's set to nil.
```
Comment (tsenart) [inline cmd/repo-updater/repos/syncer.go]: Perhaps worth mentioning in this comment that because we don't persist these errors to Postgres, there's a chance that admins will see inconsistent / flickering information in the status panel if repo-updater dies (and that that's OK for now).

File: cmd/repo-updater/repos/syncer.go [hunk 21]
```diff
@@ -311,13 +318,47 @@ func (s *Syncer) sourced(ctx context.Context) ([]*Repo, error) {
 
 	srcs, err := s.sourcer(svcs...)
 	if err != nil {
+		// Only reset source error if it was non-nil. Because if `ListRepos`
+		// produces an error through multiple syncs, the state of
+		// multiSourceErr would flip-flop between here and the `ListRepos` call
+		// further down.
```
Comment (tsenart) [inline cmd/repo-updater/repos/syncer.go]: As far as I can tell, this comment isn't consistent with what we're doing in [`setOrResetMultiSourceErr`](https://github.com/sourcegraph/sourcegraph/pull/5319/files#diff-358c3ac1d1a3c6cd1867f1808d3bdb88R344) (i.e. we're setting `s.multiSourceErr` to nil when the passed `err` is not a `*MultiSourceError`. Am I reading this wrong or are we relying on typed nil to be of the expected type?

File: cmd/repo-updater/repos/sources.go (hunk 19 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/sources.go]: Good point! That's probably not necessary anymore. I'll try to follow up on this.

File: cmd/repo-updater/repos/syncer.go (hunk 21 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: This has been adressed in the follow-up, which means this whole "setting error" mechanism (hopefully) clearer: https://github.com/sourcegraph/sourcegraph/pull/5345

File: cmd/repo-updater/repos/syncer.go (hunk 20 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/syncer.go]: will follow-up on this! 👍

File: cmd/repo-updater/repos/syncer.go (hunk 21 above)
Comment (tsenart) [reply cmd/repo-updater/repos/syncer.go]: 👍
